pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -204,58 +211,52 @@ func (s *Service) handle(msg *message) error {
 	}
-
 	logger.Infof("next valid state to transition -> %s ", next.Name())
 
-	followup, action, err := next.Execute(msg.Msg, msg.ThreadID, s.ctx)
-	if err != nil {
-		return fmt.Errorf("failed to execute state %s %w", next.Name(), err)
-	}
-	logger.Infof("follow up state that need to be executed immediately -> %s", followup.Name())
-
-	err = s.update(msg.ThreadID, next)
-	if err != nil {
-		return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
-	}
-	logger.Infof("persisted the connection using %s and updating the state to %s", msg.ThreadID, followup.Name())
+	for !isNoOp(next) {
+		logger.Infof("send pre event for state %s", next.Name())
+		// TODO change from thread id to connection id
+		// TODO pass invitation id
+		s.sendMsgEvents(&dispatcher.StateMsg{
+			Type: dispatcher.PreState, Msg: msg.Msg, StateID: next.Name(),
+			Properties: s.createEventProperties(msg.ThreadID, "")})
 
-	err = action()
-	if err != nil {
-		return fmt.Errorf("failed to execute state action %s %w", followup.Name(), err)
-	}
-	logger.Infof("finish execute state action: %s", followup.Name())
-
-	logger.Infof("send post event for state %s", next.Name())
-	s.sendMsgEvents(&dispatcher.StateMsg{Type: dispatcher.PostState, Msg: msg.Msg, StateID: next.Name()})
-
-	for ; !isNoOp(followup); followup = next {
-		logger.Infof("send pre event for state %s", followup.Name())
-		s.sendMsgEvents(&dispatcher.StateMsg{Type: dispatcher.PreState, Msg: msg.Msg, StateID: followup.Name()})
-		logger.Infof("execute next state: %s", followup.Name())
+		logger.Infof("execute next state: %s", next.Name())
 		var action stateAction
-		next, action, err = followup.Execute(msg.Msg, msg.ThreadID, s.ctx)
+		var followup state
+		followup, action, err = next.Execute(msg.Msg, msg.ThreadID, s.ctx)
 		if err != nil {
-			return fmt.Errorf("failed to execute state %s %w", followup.Name(), err)
+			return fmt.Errorf("failed to execute state %s %w", next.Name(), err)
 		}
-		logger.Infof("finish execute next state: %s", followup.Name())
-		err = s.update(msg.ThreadID, followup)
-		if err != nil {
-			return fmt.Errorf("failed to persist state %s %w", followup.Name(), err)
+		logger.Infof("finish execute next state: %s", next.Name())
+		if err = s.update(msg.ThreadID, next); err != nil {
+			return fmt.Errorf("failed to persist state %s %w", next.Name(), err)
 		}
-		logger.Infof("persisted the connection using %s and updating the state to %s", msg.ThreadID, followup.Name())
-		err = action()
-		if err != nil {
-			return fmt.Errorf("failed to execute state action %s %w", followup.Name(), err)
+		logger.Infof("persisted the connection using %s and updating the state to %s",
+			msg.ThreadID, next.Name())
+		if err := action(); err != nil {
+			return fmt.Errorf("failed to execute state action %s %w", next.Name(), err)
 		}
-		logger.Infof("finish execute state action: %s", followup.Name())
+		logger.Infof("finish execute state action: %s", next.Name())
```

Comment:
there are too many individual log messages.

Reply:
We will change all them to debug after we complete the 0.1.0 demo

Reply:
Even when you change to debug, you should have two logger statements in a row ;).

Reply:
@troyronda I change it